### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by timocov** [tradingview/lightweight-charts#362]
If you like (or not) Lightweight Charts or just have something to say about it, please leave your feedback here.

**Comment by deman4ik** [comment]
Great Charts!

Features requests:
- Add optional "price" parameter for setMarkers position inBar mode to place marker at exact place on bar (not only in the middle of the bar)
- Link several markers with lines for example for linking position's entry and exit markers
- Mouseover tooltip on Price Line axis label

**Comment by nikitamarcius** [comment]
Hi! 

I want to see much more than there is now. I understand that **lightweight-charts** is an alternative to private lib **charting library**. 

**pros**

- lightweight library
- great api with quality code
- different types of graphs for different tasks
- excellent pixel perfect


**cons**

- indicators (MACD, MA, 4MA, BB, RSI, STOCH and some more most popular)
- tool for drawing (base tool). line, Fibonacci grid, horizontal line

I wrote this in cons because there are open-source competitors that give more opportunities (indicators, tool for drawing and more). At the same time, they also have their own shortcomings, with a more comprehensive solution. 

Can share links in Discord. If you can adapt quickly, it will be awesome.

**Comment by timocov** [comment]
@deman4ik:

> Add optional "price" parameter for setMarkers position inBar mode to place marker at exact place on bar (not only in the middle of the bar)

#127

> Link several markers with lines for example for linking position's entry and exit markers

What do you mean?

> Mouseover tooltip on Price Line axis label

What kind of tooltip are you looking for? 🤔 

@nikitamarcius:

> indicators (MACD, MA, 4MA, BB, RSI, STOCH and some more most popular)

Afaik we don't plan to add indicators itself, but we've thought about adding some extra package with indicators (after #290 it'll be possible to put several series on the same price axis).

> tool for drawing (base tool). line, Fibonacci grid, horizontal line

Yeeeah, I'm afraid that adding this will add a ton of additional code... We've thoughts about allowing to write custom painters so you'll be able to add your custom tools on the chart, but it's just thoughts and I don't think that it could be done in near future.

> Can share links in Discord

Sure!

**Comment by deman4ik** [comment]
@timocov 
![Screenshot 2020-03-21 at 23 58 41](https://user-images.githubusercontent.com/913027/77236483-0d603580-6bd0-11ea-9120-ac761a90236f.png)

Example 2 circles at exact place (based on price) on bar for position entry and exit connected with line

![Screenshot 2020-03-21 at 23 59 04](https://user-images.githubusercontent.com/913027/77236510-50baa400-6bd0-11ea-8b94-f0b1df5804f4.png)

Example mouseover tooltip on Price Line axis label

**Comment by qy20092011** [comment]
Hi, what should i do if I want to know which problems are solved in the new version(1.2.2 or 2.0.0). For example, maybe the version supported it. [#129](https://github.com/tradingview/lightweight-charts/issues/129)

**Comment by AoDev** [comment]
__This library is really good for our product__, and we thank you for all the work. The lightweight side is important as a lot of our users will come from mobile. But, as already stated above, we miss some features such as an easy way to add basic indicators like MA's or RSI' or to customize the chart a bit more. Thus, in the future we might need to switch to another library.

I think, the ideal plan would be to develop a __plugin system__ so that the core keeps being lightweight while people can just add stuff like indicators optionally.

**Comment by danger89** [comment]
Thanks for sharing this piece of code and make it open-source, if you want to compete with D3 and D3FC however you need to give a bit more features out of the box I think.

As mentioned before, indeed all these basic indicators like MACD/RSI/BB/EMA/MA makes so munch sense to be added to this library. :+1: In fact, you do already have quite some advanced features, like the real-time emulation chart for example. :open_mouth: 

See your competition:
- https://d3fc.io/
- https://observablehq.com/@d3/gallery

Nevertheless, keep up the good work! I love Tradingview a lot, making these kind of libraries open-source add only more love and appreciation!!! :love_letter:

**Comment by timocov** [comment]
> Too bad you took tradingview/charting_library down.

@danger89 we don't. It's still alive, we maintain it as well and you can use it.

**Comment by danger89** [comment]
> 
> 
> > Too bad you took tradingview/charting_library down.
> 
> @danger89 we don't. It's still alive, we maintain it as well and you can use it.

Well it's a 404 for me. I know you can download it, but I mean it's not open-source.

![image](https://user-images.githubusercontent.com/628926/79867458-a60bf000-83de-11ea-8015-ec87b6ae5b08.png)

**Comment by allartdo** [comment]
Copyd from: https://github.com/tradingview/lightweight-charts/issues/412 with a request from: @timocov 

Alright, I probably change library since this isn't my only problem. I will give you some critical feedback why I get kinda triggered because of the Lightweight chart library:

Documentation is kinda hard to follow. Its diffrent then the usuall documentation like, chart.js, leaflet etc. It took for me a lot of time to find out how to get to all parts of the documentation because if I follow the /docs/readme I see only 3 extra links to move to any after, provide a list with all documentation pages and u already make it much better. For beginners it might take infinate now and experts may not even start using this.

While using the library it feels like there are just missing parts, getPrice(), getMarkers, setMarker (instead of setMarkers), removeMarker(), removeAllMarkers() etc. etc... I understand this is open source and you need to program most of it by urself but these functions seem pritty much standard, in a library like this, for me.

If I look at the #195 link you send and see that that guy posted last year september and again another guy did in februari and ended in march and now me with still no answer/commit or at least pseudo code, I would say you must find a real fan of your library to get it fixed. So your last answer feels pritty much a useless/duplicate and a "politic correct" answer if u understand what im saying. My answer would be (example without actuall fact of u or ur team): Yes allartdo, since we are not planned to fix this and our goal doesn't lies here, you really have to build this in urself or have to wait untill someone else does. If you do want to try and fix it you might want to look in this or these file('s). This might shorten your time searching for the correct files. Also these function may help you solve your problem: ...

I hope you understand what im trying to do, I make sure I dont need to program it myself but give the complaining guy with an issue a big advantage to actually make it himself. Just like in an shop, the customer wants a mars chocelate bar but since the shop doesn't have it the shop say's sorry we dont have it and offers a kitkat chocelate bar. This to make the end result of the customer hopefully happy (and get more money). Again both parties end up with a bigger change of being happy.

As I said in the beginning of my question, I posted this question on stackoverflow (and other questions about the library) and it seems there are not that many people using the library in Javascript. This is, ofcourse, not a fact but a feeling I get, which could be wrong.

If I read the first lines in the readme here: https://github.com/tradingview/lightweight-charts then it seems to me that you focus mostly on performance and speed. I understand you dont want to have a chart that laggs, but with the new tech today and very strong processors you would need to really fk up ur code to make it this laggy. If your pc or device is so bad that it can't handle the chart it must be a pc of at least 10 years old in my opinion. I would say if my goal would be userbility to defeat library's like Chart.js. Since tradingview.com is one of the most used charting viewers to quikly edit data. Example of userbility: Be able to simply rebuild one of your (in web) made charts to a programmed chart. This is, in my opinion, again very basic stuff.

Again my intention is not to break a hart but to make it stronger, I give feedback which may be hard but nice to have.
https://www.google.com/search?q=feedback+quote&sxsrf=ALeKk03_-jy2bp_6XgeT26HewGWmMclrMA:1587561885840&source=lnms&tbm=isch&sa=X&ved=2ahUKEwiNtY7ykPzoAhWP5KQKHc_fA5QQ_AUoAXoECA0QAw&biw=1920&bih=937

Use it in advance :)

**Comment by timocov** [comment]
@allartdo thanks for such critical feedback. We do really appreciate every feedback. This days seems everyone trying to be gently and avoid writing such feedbacks to don't offence nobody (IMO).

> Documentation is kinda hard to follow. Its diffrent then the usuall documentation like, chart.js, leaflet etc. It took for me a lot of time to find out how to get to all parts of the documentation because if I follow the /docs/readme I see only 3 extra links to move to any after, provide a list with all documentation pages and u already make it much better. For beginners it might take infinate now and experts may not even start using this.

I agree with that. I admit that our docs aren't the best and there are a ton of things that could be improved to make them better, improve readability and so on.

> While using the library it feels like there are just missing parts, getPrice(), getMarkers, setMarker (instead of setMarkers), removeMarker(), removeAllMarkers() etc. etc... I understand this is open source and you need to program most of it by urself but these functions seem pritty much standard, in a library like this, for me.

I'm sure that no one just didn't think about it then it could be useful for someone. The part with providing/getting data is kind of tricky. I meant you are that guy who provides that data to us, why are you asking for them? The data you passed to the library is processed in different ways internally and might be changed in one way without ability to restore the original data, just because it's to have the simplest way to provide data in API, but store that data internally optimized for internal purposes (speed, performance, access, etc). If you really need this, please file an issue with your suggestion and I think we can continue to discuss it there (please provide use cases as well). I should say that almost all new features (which come after 1.0 release) have or require to have proposal posted in the issue so everyone can see it and provide it's feedback. Markers was introduced in #24, and it seems nobody asked for things you've written in your message. Thus I'd say that no one wants/needs it except you. But anyway, we're open for any suggestion which can make our API better.

> Yes allartdo, since we are not planned to fix this and our goal doesn't lies here, you really have to build this in urself or have to wait untill someone else does. If you do want to try and fix it you might want to look in this or these file('s). This might shorten your time searching for the correct files. Also these function may help you solve your problem: ...

If we don't plan to have some feature/issue in lightweight-charts, we'll say that in the issue. We can't just do every feature in one time/release, only critical/major/business required features (because it's our employer and it pays money for that basically) because we don't work at the project full-time and have really limited resources. That's one of reason why we decided to make it open-sourced - even we haven't enough time to do everything, there are a lot of developers, who can spend a bit of it's time to implement feature/fix bug/help with docs/even to make proposal and etc.

> If I look at the #195 link you send and see that that guy posted last year september and again another guy did in februari and ended in march and now me with still no answer/commit or at least pseudo code, I would say you must find a real fan of your library to get it fixed. So your last answer feels pritty much a useless/duplicate and a "politic correct" answer if u understand what im saying

So, regard this one - we do really think that feature might be useful (but see previous paragraph). But before implementing it, we need to have a proposal of changes (that's why there is a label "need proposal"), so we can see what changes we'll have, discuss it etc. If you don't want to spend your time to improve the project with your code, but you understand how it should be looked in the API and you're ready to share your idea with the community - please do it. "need proposal" doesn't mean that we as maintainers should provide it. This means that we need to think about this feature and provide what API changes we need to have here to make it work.

> I hope you understand what im trying to do, I make sure I dont need to program it myself but give the complaining guy with an issue a big advantage to actually make it himself. Just like in an shop, the customer wants a mars chocelate bar but since the shop doesn't have it the shop say's sorry we dont have it and offers a kitkat chocelate bar. This to make the end result of the customer hopefully happy (and get more money). Again both parties end up with a bigger change of being happy.

I'm not sure that this comparison is applicable to open-source projects (which doesn't earn any money from that). Can you elaborate it a bit more please? Does this comment related to the previous paragraph with #195?

> As I said in the beginning of my question, I posted this question on stackoverflow (and other questions about the library) and it seems there are not that many people using the library in Javascript. This is, ofcourse, not a fact but a feeling I get, which could be wrong.

Looks like we don't monitor questions there. I'll up this question with the team. The library's life started in less than 1 year, so I can understand why it's not so popular though yet (also there is a point that the aim of the library is financial charts, which have it's niche and it's not React*/Angular/vuejs which could be used in almost any project).

> If I read the first lines in the readme here: tradingview/lightweight-charts then it seems to me that you focus mostly on performance and speed. I understand you dont want to have a chart that laggs, but with the new tech today and very strong processors you would need to really fk up ur code to make it this laggy. If your pc or device is so bad that it can't handle the chart it must be a pc of at least 10 years old in my opinion.

I strongly disagree with you, as do a lot of people I know personally/read in twitter/etc. Why did you talk about PCs, but didn't talk about mobile devices? It's super easy to make your application slow with just several lines of code in the way your super-faster PC won't handle it. We, as the library devs, just can't make it slow, because on the page where the library will be used might be a lot of other libraries/frameworks which in a couple affects the page's performance. If we'll slow, nobody will use the library, because it makes its site slow (don't forget about mobile devices and people outside of your country/city where you have a stable 4G internet connection).

> I would say if my goal would be userbility to defeat library's like Chart.js. Since tradingview.com is one of the most used charting viewers to quikly edit data. Example of userbility: Be able to simply rebuild one of your (in web) made charts to a programmed chart. This is, in my opinion, again very basic stuff.

Not sure that I understand your point. Could you elaborate it please?

Overall, thank you so much for your feedback! The only feedbacks help us do our best to improve the library in any way. We admit that documentation might be one of our weakest things, so we do really appreciate any help with that. Thus, if anyone would like to help us with that (we can even consider to have full or part-time paid support/work with that). Also, I should say that we do our best here and really open to any kind of constructive discussion ❤️.

**Comment by leeoniya** [comment]
> I understand you dont want to have a chart that laggs, but with the new tech today and very strong processors you would need to really fk up ur code to make it this laggy.

this is completely false. a chart with 1,000 datapoints is not very large. and it's easy to have 10 charts with 100 datapoints or more on one page. for example: https://arewefastyet.com/win10/overview?numDays=60. rendered with a fast library, that page would load in a couple hundred ms. i have an issue open to improve exactly this: https://github.com/mozilla-frontend-infra/firefox-performance-dashboards/issues/293. there are plenty of uses where the vast majority of JS charting libs do nothing but waste CPU cycles.

**Comment by leeoniya** [comment]
if anyone is interested in how lightweight-charts performs vs many other charting libs, it's now included in uPlot's benchmarks [1]. @timocov, if there are any perf improvements in the future, definitely ping me so i can update the numbers.

[1] https://github.com/leeoniya/uPlot#performance

**Comment by allartdo** [comment]
@leeoniya I dont really understand now why you belive its better to have performance as main goal. What is your target audience? People with slow computers/devices? What does it bring you in the end? Do you get more users this way? If so why does it work? Im searching for a bit more information.

I think at this moment I may learn even more then you guys do because of what I say. I see that I have said things that aren't completely realistic and some even disagree. But for me this does not count like a failure. I like to fall down and stand up, even on purpose if I have to. I am a student in The Netherlands and have been at school learning programming for almost three years now. Back then in the beginning I did not know how to write any code and did not know what HTML, CSS or Javascript was. So you could understand now that I am still not a very good programmer and I am not very experienced. Knowing this, its not this bad for me to go down this way, I already learned a lot from you guys.

@timocov 

> @allartdo thanks for such critical feedback. We do really appreciate every feedback. This days seems everyone trying to be gently and avoid writing such feedbacks to don't offence nobody (IMO).

You see this in the office/reallife or in this online way of speaking? Or both ofcourse :)

> I'm sure that no one just didn't think about it then it could be useful for someone.

I get the feeling I don't just dont know why people start using this library. Right now I think people use it because they can automate trading bitcoin or other coins. And I would never start instandly building/using live data and start trading. I would like a testcase first, to make sure I get profit or not. Being able to get the price easily would help in this.

> If we don't plan to have some feature/issue in lightweight-charts, we'll say that in the issue. We can't just do every feature in one time/release, only critical/major/business required features (because it's our employer and it pays money for that basically) because we don't work at the project full-time and have really limited resources. That's one of reason why we decided to make it open-sourced - even we haven't enough time to do everything, there are a lot of developers, who can spend a bit of it's time to implement feature/fix bug/help with docs/even to make proposal and etc.

Understandable 👍 

> Does this comment related to the previous paragraph with #195?

Yes, I see now that I may have made things more unclear by saying that... I will leave that for what it is if you dont mind.

> Looks like we don't monitor questions there. I'll up this question with the team.

Cool!

> The library's life started in less than 1 year, so I can understand why it's not so popular though yet.

I understand too now 👍 

> I strongly disagree with you, as do a lot of people I know personally/read in twitter/etc. Why did you talk about PCs, but didn't talk about mobile devices?

>  If your pc or device is so bad that it can't handle the chart it must be a pc of at least 10 years old in my opinion.

Device could by anything in the way I said this. And also is how I meant it

> It's super easy to make your application slow with just several lines of code in the way your super-faster PC won't handle it. We, as the library devs, just can't make it slow, because on the page where the library will be used might be a lot of other libraries/frameworks which in a couple affects the page's performance. If we'll slow, nobody will use the library, because it makes its site slow (don't forget about mobile devices and people outside of your country/city where you have a stable 4G internet connection).

I dont want you to make it slow. I hoped that small function (here it comes again :) like getPrice() would not make it slow. I can understand that when you have 100 or 1000 of these functions it will get slow. I agree and understand what you say. I think I do not have enough experience to know how this works perfectly. PS: Check out [Starlink](https://www.google.com/search?q=starlink&oq=starlink&aqs=chrome..69i57j69i59l3j0l3j69i60.3096j0j7&sourceid=chrome&ie=UTF-8). Soon bad internet connections might be gone :P

> Not sure that I understand your point. Could you elaborate it please?

I am wondering which target group you and your team has. Traders? Ordinary people with basic internet (hehe)? Or maybe company's that want to show there financials?

> Overall, thank you so much for your feedback! The only feedbacks help us do our best to improve the library in any way. We admit that documentation might be one of our weakest things, so we do really appreciate any help with that.

Happy to hear this! And thank you and others too ofcourse for helping me and making me smarter :) 👍 

> Thus, if anyone would like to help us with that (we can even consider to have full or part-time paid support/work with that). Also, I should say that we do our best here and really open to any kind of constructive discussion ❤️.

Is there a website to get this information? Like amount of payment, to-do's, how you guys would like to have it, how to join/apply etc.

**Comment by timocov** [comment]
@allartdo 

> Device could by anything in the way I said this. And also is how I meant it

Ok, we got your point. We have another way of looking at things but I don't think that it's what we should discuss here. If we'd prefer to aim on performance and small size - it's our choice and we want to make it so. We have 10M+ users a month and we have knowledge that performance is critical at least for us, that's why we always trying to do our best with that. If you don't prefer that - that's okay and we don't argue with you here and don't try to change your mind.

> PS: Check out Starlink. Soon bad internet connections might be gone :P

Let's talk when it happen 😉

> Is there a website to get this information? Like amount of payment, to-do's, how you guys would like to have it, how to join/apply etc.

This is uncommon thing and I'm not sure we have active vacancy for that anywhere. If you'd like to continue with that - please email me etimokhov@tradingview.com with what you'd like to do and your expectation - I'll forward to proper guy.

**Comment by costaluu** [comment]
Hello, I've been following this project for some time and really find it the most promising in the segment. 

Speaking about the issue of new features, many need a change proposal and I understand your developers' vision, but currently the documentation scenario of the code is very restricted which makes it very difficult for anyone to take the initiative and not everyone has time to sit down and understand all the nuances of the source code.
In most of this quarantine I tried to understand the code and even managed to do something haha, but soon I gave up. Today I lack enough time to do my own tasks but I believe there are several people willing to contribute to this project.
I'm here to recommend that this work of creating a documentation of all the code's API be done by you developers making it easier for people to contribute more

So that's it, thanks ;)

**Comment by jazoom** [comment]
This is a great library, though I agree the documentation is still a work-in-progress.

How do I get the price/date coordinates under the cursor/crosshair? I know via the `series` object you can get the date of the candle, so I guess that's half of it. But what about price level? I don't mean the OHLC of the candlestick. I mean if the crosshair is hovering above a candle at a price of 1000 (candle high is 700) then I want the coordinate 1000.

Example usage: stick a `priceLine` at the location of the crosshair on click.

```
chart.subscribeClick(param => {
  series.createPriceLine({
    price: <HOW_TO_GET_THIS>
  });
});
```

**Comment by wtlin1228** [comment]
Is this [site](https://www.futures-ai.com/quote) using the lightweight-charts?

It looks like the full version of trading view as the tradingview.com provides. I want to know where can I get the source code.

Is anybody can help me?

**Comment by allartdo** [comment]
@wtlin1228 Not sure if this is the place to ask your question but here is your answer:

It is indeed a version of tradingview.com, if its the full version I dont know but they definitly get the most out of it. At the [official tradingview website](https://www.tradingview.com/HTML5-stock-forex-bitcoin-charting-library/) you can choose what library you want. I use Lightweight Charts since its free without needing an company name etc. You should really watch a video or search on internet to see what suits you best. Good luck
